### ISSUE TITLE
support regular expression ??

### ISSUE THREAD

**Issue by looker53** [jmespath/jmespath.py#213]
e.g.
```
book.p[content =~ /^pica+$/i]
```

**Comment by simonw** [comment]
I don't think regular expressions should be part of the core JMESPath syntax because different programming languages have different regular expression implementations, so it wouldn't be possible to guarantee that a regular expression run using JMESPath on Python would have the exact same results on PHP or Ruby or Lua or Go.

Instead I suggest writing your own custom function that makes Python regular expressions available to your code: https://github.com/jmespath/jmespath.py/blob/0.10.0/README.rst#custom-functions

**Comment by notjames** [comment]
> I don't think regular expressions should be part of the core JMESPath syntax because different programming languages have different regular expression implementations, so it wouldn't be possible to guarantee that a regular expression run using JMESPath on Python would have the exact same results on PHP or Ruby or Lua or Go.
> 
> Instead I suggest writing your own custom function that makes Python regular expressions available to your code: https://github.com/jmespath/jmespath.py/blob/0.10.0/README.rst#custom-functions

I completely disagree with this assertion. Though many languages do indeed have different implementations, I estimate 99.999% of them use a PCRE library/implementation for regexes as PCRE has become the defacto standard for regular expressions. Moreover, if every language used the rationale to defer to implement regular expressions because other languages do it differently then no language would implement them. I find myself deferring to `jq` 90% of the time because `jp` doesn't employ regular expressions.

**Comment by mswezey23** [comment]
I'd be in huge support for adding regex support!

**Comment by simonw** [comment]
There's a useful comparison table of the differences between different language implementations here: https://web.archive.org/web/20130830063653/http://www.regular-expressions.info:80/refflavors.html

**Comment by simonw** [comment]
`jq` is an interesting comparison here. It actually bundles the full implementation of the regular expression library that it uses in the C implementation of the tool - then different programming languages have bindings that depend on that shared implementation: https://github.com/stedolan/jq/wiki/FAQ#language-bindings

`jmespath` works differently: rather than shipping a single C implementation that is shared by different languages, it encourages entirely fresh implementations in different languages with no shared code, but sharing instead a thorough set of compliance tests: https://jmespath.org/libraries.html

This approach could be used to add regular expression support, by spinning up a thorough compliance test and maybe defining a subset of regular expression syntax that must be supported in order for an implementation to pass the test.

It would be a lot of work though!

**Comment by simonw** [comment]
Here's a really interesting relevant proposal, for an interoperable subset of regular expressions: https://cabo.github.io/iregexp/draft-bormann-jsonpath-iregexp.html
